code/go/0chain.net/chaincore/chain/worker.go

```diff
@@ -370,7 +370,7 @@ func (c *Chain) finalizeBlockProcess(ctx context.Context, fb *block.Block, bsh B
 		}
 
 		prevBlockHash := pr.GetBlockHash()
-		if prevBlockHash == "" {
+		if prevBlockHash == "" || !pr.IsFinalized() {
```

Comment:
previous round is not always a finalized round, if we had more than one finalizations on the previous round, then it can't be finalized and rounds should be finalized by chain later on, when single distinct round will be chosen, it will be done in method `c.finalizeBlock(ctx, fb, bsh)`

Reply:
I think what you mean is the [chain.FinalizeRound](https://github.com/0chain/0chain/blob/d6cb4dc775b7cdd9ec2f78920d04e18148ad3167/code/go/0chain.net/chaincore/chain/protocol_round_main.go#L10), we run it whenever a block is notarized, and we don't require its previous round is finalized. But this finalize process is the last step to finalize a round, there's no later on actually, on sharders, its previous round must have been finalized, otherwise, there will have FB gaps.

Reply:
ah you mean that we should have already processed this round before, just when when were processing frchain here https://github.com/0chain/0chain/blob/81ba565e6e5e6597e94c08bbe68c9093d771ec48/code/go/0chain.net/chaincore/chain/protocol_round.go#L325

Reply:
yes